src/lib/transition_frontier/transition_frontier.ml

```diff
@@ -1 +1,274 @@
+           (Blockchain_state.ledger_hash blockchain_state)) ) ;
+    let mask = Merkle_ledger.derive ledger in
+    Merkle_mask.apply mask (External_transition.ledger_builder_diff root) ;
+    assert (
+      Ledger_hash.equal (Merkle_mask.root mask)
+        (Ledger_builder_hash.ledger_hash
+           (Blockchain_state.ledger_builder_hash blockchain_state)) ) ;
+    let root_hash = Protocol_state.hash protocol_state in
+    let root_breadcrumb = {Breadcrumb.transition= root; mask} in
+    let root_node =
+      {breadcrumb= root_breadcrumb; successor_hashes= []; length= 0}
+    in
+    let table = State_hash.Table.of_alist_exn [(root_hash, root_node)] in
+    {root_ledger= ledger; root= root_hash; best_tip= root_hash; table}
+
+  let find t hash =
+    let open Option.Let_syntax in
+    let%map node = Hashtbl.find t.table hash in
+    node.breadcrumb
+
+  let find_exn t hash =
+    let node = Hashtbl.find_exn t.table hash in
+    node.breadcrumb
+
+  let path t breadcrumb =
+    let rec find_path b =
+      let hash = Breadcrumb.hash b in
+      let parent_hash = Breadcrumb.parent_hash b in
+      if State_hash.equal parent_hash t.root then [hash]
+      else hash :: find_path (find_exn t parent_hash)
+    in
+    List.rev (find_path breadcrumb)
+
+  let iter t ~f = Hashtbl.iter t.table ~f:(fun n -> f n.breadcrumb)
+
+  let root t = find_exn t t.root
+
+  let best_tip t = find_exn t t.best_tip
+
+  let successor_hashes t hash =
+    let node = Hashtbl.find_exn t.table hash in
+    node.successor_hashes
+
+  let rec successor_hashes_rec t hash =
+    List.concat
+      (List.map (successor_hashes t hash) ~f:(fun succ_hash ->
+           succ_hash :: successor_hashes_rec t succ_hash ))
+
+  let successors t breadcrumb =
+    List.map (successor_hashes t (Breadcrumb.hash breadcrumb)) ~f:(find_exn t)
+
+  let rec successors_rec t breadcrumb =
+    List.concat
+      (List.map (successors t breadcrumb) ~f:(fun succ ->
+           succ :: successors_rec t succ ))
+
+  (* Adding a transition to the transition frontier is broken into the following steps:
+   *   1) create a new node for a transition
+   *     a) derive a new mask from the parent mask
+   *     b) apply the ledger builder diff to the new mask
+   *     c) form the breadcrumb and node records
+   *   2) add the node to the table
+   *   3) add the successor_hashes entry to the parent node
+   *   4) move the root if the path to the new node is longer than the max length
+   *     a) calculate the distance from the new node to the parent
+   *     b) if the distance is greater than the max length:
+   *       I  ) find the immediate successor of the old root in the path to the
+   *            longest node and make it the new root
+   *       II ) find all successors of the other immediate successors of the old root
+   *       III) remove the old root and all of the nodes found in (II) from the table
+   *       IV ) merge the old root's merkle mask into the root ledger
```

Comment:
this will trigger an upwards propagation of changes in all masks in the system (and this will be probably too slow to be synchronous no?)

Reply:
We might have to do something crazy with dirty bits or something to lazily propagate the changes

Reply:
Discussed in person.

Reply:
See #1177 , mask merging is cheap!